File: packages/go/internal/seguranca/ataque.go
package seguranca

// Ataque é o contrato dos ataques da suíte.
type Ataque interface {
	Nome() string

	// Executar roda o ataque contra o alvo e devolve o resultado.
	// Deve devolver um *SkipAtaqueException se o alvo não suportar os recursos
	// necessários (ex: não expõe GerarKeystreamBruto).
	Executar(alvo AlvoCriptografico) (ResultadoAtaque, error)
}

File: packages/rust/src/seguranca/ataque.rs
//! Contrato dos ataques.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::resultado_ataque::ResultadoAtaque;

/// Erro de execução de um ataque.
///
/// `Skip` corresponde à `SkipAtaqueException` (o alvo não suporta o que o
/// ataque precisa). `Falha` corresponde a qualquer exceção inesperada.
#[derive(Debug, Clone)]
pub enum ErroAtaque {
    Skip(String),
    Falha(String),
}

pub trait Ataque {
    fn nome(&self) -> String;

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque>;
}
